src/client.ts

```diff
@@ -30,56 +24,39 @@ function eosPromisified(stream: NodeJS.ReadableStream | NodeJS.WritableStream):
 }
 
 export class MqttClient extends EventEmitter {
-  _incomingPacketParser: MqttParser;
-  _options: ConnectOptions;
-  disconnecting: any;
-  connecting: boolean;
-  connected: boolean;
-  errored: boolean;
-  _eos: Promise<void> | undefined;
-  conn: Duplex | Socket;
-  _clientLogger: Logger;
-  /**
-   * Use packet ID as key if there is one (e.g., SUBACK)
-   * Use packet type as key if there is no packet ID (e.g., CONNACK)
-   */
-  // TODO: This should be removed after we remove CONNECT into the sequencer
-  _inflightPackets: Map<string | number, (err: Error | null, packet: IPacket) => void>;
-  private _packetSequencer = new sequencer.MqttPacketSequencer(write.bind(null, this));
-
-  constructor(options: ConnectOptions) {
+  private _incomingPacketParser: MqttParser;
+  private _options: ClientOptions;
+  private _disconnecting: boolean;
+  private _connecting: boolean;
+  private _connected: boolean;
+  private _errored: boolean;
+  private _eos: Promise<void> | undefined;
+  private _conn: Duplex | Socket;
+  private _clientLogger: Logger;
+  private _packetSequencer: MqttPacketSequencer;
+
+  constructor(options: ClientOptions) {
     super();
     // assume the options have been validated before instantiating the client.
-    this.connecting = false;
-    this.connected = false;
-    this.errored = false;
-    this.disconnecting = false;
-    this._inflightPackets = new Map();
-
-    // Using this method to clean up the constructor to do options handling
-    logger.trace(`populating internal client options object...`);
-    this._options = {
-      clientId: defaultClientId(),
-      ...defaultConnectOptions,
-      ...options,
-    };
+    this._options = options;
+    this._connecting = false;
+    this._connected = false;
+    this._errored = false;
+    this._disconnecting = false;
+    this._packetSequencer = new MqttPacketSequencer(this._options, write.bind(null, this));
 
     this._clientLogger = logger.child({ id: this._options.clientId });
 
-    this.conn = this._options.customStreamFactory
+    this._conn = this._options.customStreamFactory
       ? this._options.customStreamFactory(this._options)
       : connectionFactory(this._options);
 
     // many drain listeners are needed for qos 1 callbacks if the connection is intermittent
-    this.conn.setMaxListeners(1000);
+    this._conn.setMaxListeners(1000);
 
     this._incomingPacketParser = mqttParser(this._options);
 
-    // Handle incoming packets this are parsed
-    // NOTE: This is only handling incoming packets from the
-    // readable stream of the conn stream.
-    // we need to make sure that the function called on 'packet' is bound to the context of 'MQTTClient'
-    this._incomingPacketParser.on('packet', this.handleIncomingPacket.bind(this));
+    this._incomingPacketParser.on('packet', this._packetSequencer.handleIncomingPacket.bind(this._packetSequencer));
```

Comment:
maybe dumb question: is handleIncomingPacket bound to _packetSequencer because the context for packets should be _packetSequencer's context?

Reply:
It's because the method `MqttPacketSequencer.handleIncomingPacket()` should have the context of `this._packetSequencer`